### PR TITLE
task: Remove support for ruby 2.4 and rails 4.2

### PR DESCRIPTION
**Pull request by KapilSachdev**
- Remove obsolete rails version helpers for rails < 5
- Update all gem versions
- Disable Naming/VariableNumber Cop
- Use bionic distribution for CI

### CODE REVIEW

**Comment by KapilSachdev** [comment]
Travis CI builds have become [very slow and are delayed](https://travis-ci.community/t/build-delays-for-open-source-project/10272) as well due to [recent changes](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing) in pricing structure which will ultimately lead to a credit based system for OSS as well.
@mcmire  What do you think about switching to some other build platforms like [buildkite](https://buildkite.com/pricing) offers unlimited as of now

**Comment by mcmire** [comment]
@KapilSachdev Ah, I didn't know about those changes. That's too bad.

We use Buildkite at my company, so I'm at least roughly familiar with it. I'm not 100% sure on how it was set up, but I know it's worked really well for us and is super customizable and fairly easy to set up. The only part I'm not sure about is how to emulate Travis's matrix, although I'm sure someone's written a blog post about it. But yeah... based on what people are saying in the Travis Forums it sounds like this is something we would need to do before the end of the year. :(

**Comment by KapilSachdev** [comment]
Yes, we will need to do that soon.
I will try Buildkite with my fork and update the findings here.

**Comment by KapilSachdev** [comment]
Before starting on `buildkite`, I gave github actions a try.
It is working fine with `spec:unit` but `spec:acceptance` are [failing](https://github.com/KapilSachdev/shoulda-matchers/runs/1435714749).

**Comment by VSPPedro** [comment]
Hi, @KapilSachdev and @mcmire,  I found this [Drop legacy callback matchers under Rails 5](https://github.com/thoughtbot/shoulda-matchers/commit/6ea9afc1066502fdd5c556fa9edd4ce7085169fc) commit and I'm just sharing with you because I think we need to do that after dropping the support for 4.2 and before creating a major release.

It's just to remind us. Maybe an issue should be a better way to keep this registered. :smile: 

I will create one later, if you think the same way, of course.

File: spec/support/unit/helpers/active_record_versions.rb [hunk 1]
```diff
@@ -11,42 +11,6 @@ def active_record_version
       Tests::Version.new(::ActiveRecord::VERSION::STRING)
     end
 
-    def active_record_enum_supports_prefix_and_suffix?
```
Comment (KapilSachdev) [inline spec/support/unit/helpers/active_record_versions.rb:14]: @VSPPedro Is this what you are referring to when you commented about legacy controller callbacks?

I think removing all the legacy callbacks under rails < 5 in this PR would be good.

**Comment by mcmire** [comment]
Yes, going off of what @VSPPedro is saying, if we are ready to work on 5.0.0 then that's totally fine to upgrade to Rails 6.1 at the same time as removing Rails 4.2 etc. My understanding however was that we were releasing a minor version next, is that true? In any case it might be more useful to split this into two PRs — one that adds Rails 6.1, another that removes 4.2. That way if we want to add Rails 6.1 now we can do that without having to remove 4.2 yet.

**Comment by KapilSachdev** [comment]
>  if we are ready to work on 5.0.0 then that's totally fine to upgrade to Rails 6.1 at the same time as removing Rails 4.2 etc

Yes, The PR was created focusing the next major release i.e. `5.0`, which would change the minimum ruby and rails version support to `2.5` and `5.0` respectively.
I think we are ready. Is there any roadmap you have in mind for `5.0` @mcmire .

>My understanding however was that we were releasing a minor version next, is that true?

If your understanding arises due to the [`4.5.0`](https://github.com/thoughtbot/shoulda-matchers/milestone/20) milestone, I just created it thinking the same.
@mcmire  What do you think about skipping a minor one and making a major release?

> In any case it might be more useful to split this into two PRs — one that adds Rails 6.1, another that removes 4.2

Sounds good. Will split it into two.

> That way if we want to add Rails 6.1 now we can do that without having to remove 4.2 yet.

This 6.1 is a release candidate, and i'm not sure about the stable release timeline. So I think adding it with  `shoulda-matchers ~4` might delay `~5`.

**Comment by mcmire** [comment]
@KapilSachdev Okay, good. I figured this was for 5.0.0 but I wasn't sure.

Speaking of 5.0.0, I just wrote up my thoughts about it including possible backward-incompatible changes we could make, so I'd be curious to hear your thoughts here: #1380

**Comment by KapilSachdev** [comment]
> I figured this was for 5.0.0 but I wasn't sure.

My bad that I didn't added any info to the draft about it being a part of `5.0`.
I was into just testing the changes, though i found out later that travis ci is too slow to help.

> I think it may be useful to release a minor version next

Yes, definitely. I totally forgot about the [discussion](https://github.com/thoughtbot/shoulda-matchers/pull/1369#issuecomment-725643792) of creating a deprecation cycle.

File: spec/support/unit/helpers/active_record_versions.rb (hunk 1 above)
Comment (VSPPedro) [reply spec/support/unit/helpers/active_record_versions.rb:14]: I'm so sorry. I thought you hadn't made that change. I should have analyzed the changes better.

> I think removing all the legacy callbacks under rails < 5 in this PR would be good.

I agree.

**Comment by VSPPedro** [comment]
Hello, @KapilSachdev, I hope you are well. Do you still intend to work on it?

**Comment by KapilSachdev** [comment]
Hello @VSPPedro,
Yes, I will work on it and try to complete it soon.

**Comment by VSPPedro** [comment]
Hi, @KapilSachdev, if you update this with master the problem with Travis will be solved thanks to this PR: #1397.

You will still need to update the file responsible file for the ci workflow on the GithubActions. I think it's only necessary to remove the [rails_4_2](https://github.com/thoughtbot/shoulda-matchers/blob/master/.github/workflows/ci.yml#L39) appraisal and all the [content](https://github.com/thoughtbot/shoulda-matchers/blob/master/.github/workflows/ci.yml#L43-L46) of the `matrix.exclude`, but maybe I'm wrong.

I hope that helps

😅

**Comment by VSPPedro** [comment]
Hello, @KapilSachdev, I hope you are well.

This has been covered by another PR.

Thanks!

**Pull request closed without merging**